File: implementation/product/src/main/java/com/financetracker/product/logic/model/Product.java
package com.financetracker.product.logic.model;

import com.financetracker.product.infrastructure.db.converter.LabelSetConverter;
import jakarta.persistence.*;
import lombok.*;

import java.util.HashSet;
import java.util.Set;

@Builder(builderMethodName = "with")
@AllArgsConstructor
@Getter
@NoArgsConstructor
@Table(name = "product")
@Entity
public class Product {

    @Id
    private String id;
    private String name;
    private String description;

    @Embedded
    @AttributeOverride(name = "amount", column = @Column(name = "price", precision = 27, scale = 6))
    private MonetaryAmount price;

    private Category category;

    @Setter
    @Embedded
    @AttributeOverride(name = "servingSize", column = @Column(name = "nutrition_serving_size", precision = 27, scale = 6))
    @AttributeOverride(name = "calories", column = @Column(name = "nutrition_calories", precision = 27, scale = 6))
    @AttributeOverride(name = "carbohydrates", column = @Column(name = "nutrition_carbohydrates", precision = 27, scale = 6))
    @AttributeOverride(name = "protein", column = @Column(name = "nutrition_protein", precision = 27, scale = 6))
    @AttributeOverride(name = "fat", column = @Column(name = "nutrition_fat", precision = 27, scale = 6))
    @AttributeOverride(name = "sugar", column = @Column(name = "nutrition_sugar", precision = 27, scale = 6))
    private Nutrition nutrition;

    @Convert(converter = LabelSetConverter.class)
    @Column(name = "labels", nullable = false)
    protected Set<Label> labels = new HashSet<>();

    public boolean nutritionRequested() {
        return category.equals(Category.FOOD) || category.equals(Category.DRINKS);
    }
}

File: implementation/ui-financetracker/src/app/product/product.ts

export class Product {
    id: string;
    name: string;
    description: string;
    price: MonetaryAmount;
    category: Category;
    size: number;
    nutrition: Nutrition | null;
    labels: string[];

    constructor(id: string, name: string, description: string, price: MonetaryAmount, category: Category, size: number, nutrition: Nutrition, labels: string[]) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.price = price;
        this.category = category;
        this.size = size;
        this.nutrition = nutrition;
        this.labels = labels;
    }
}

export class MonetaryAmount {
    amount: number | null;

    constructor(amount: number) {
        this.amount = amount;
    }
}

export class Nutrition {
    servingSize: number;
    calories: number;
    carbohydrates: number;
    protein: number;
    fat: number;
    sugar: number;

    constructor(servingSize: number, calories: number, carbohydrates: number, protein: number, fat: number, sugar: number) {
        this.servingSize = servingSize;
        this.calories = calories;
        this.carbohydrates = carbohydrates;
        this.protein = protein;
        this.fat = fat;
        this.sugar = sugar;
    }
}

export enum Category {
    FOOD = "FOOD",
    DRINKS = "DRINKS",
    HOUSEHOLD = "HOUSEHOLD",
    ESSENTIALS = "ESSENTIALS",
    PET_SUPPLIES = "PET_SUPPLIES"
}
